Add sum-fibs program

vaibhavarora14 opened the pull request:
**Program strategy -**
- Get `fibonacci numbers` starting from 1 in for loop, and apply `addition` operation on them for `total sum`.
- When `fiboancci number` is higher than program input, for loop ends and `total sum` is returned from program

Commit:
#5 Add requirements of sum-fibs program

Commit:
#5 Add tests for sum-fibs program

Commit:
\#5 Add 3 utilities required for sum-fib program
Utilities are
- isNumber: for type checking
- fibonacciGenerator: to generate fibonacci from a starting point
- returnNumberGeneator: a cool one for me to get values in two variables with ecmascript destructuring

Commit:
#5 Add isOddNumber utility as per program specs

Commit:
Add sum-fibs implementation using added utilities

Commit:
#5 Add isPositive utility, inspired from tests

Commit:
#5 Add exception handling blocks for program safety

ArfatSalman commented on js-exercises/sum-fibs/sumFibs.js:
```diff
@@ -0,0 +1,59 @@
+
+const isPositiveNumber = (arg) => isNumber(arg) && (arg >= 0);
+
+function* returnNumberGenerator(number) {
```
Why this? Since you've used this in only one place.

vaibhavarora14 replied:
It was cool for me to use `destructuring assignment` with help of generator.
Now it's not making that much sense, and I will remove this.

vaibhavarora14 replied:
Removed with 8ab346c55474d57acbd63b0f97940c49cb53006b

ArfatSalman commented on js-exercises/sum-fibs/sumFibs.js:
```diff
@@ -0,0 +1,59 @@
+
+const isOddNumber = (arg) => isNumber(arg) && (arg % 2 !== 0);
+
+const isPositiveNumber = (arg) => isNumber(arg) && (arg >= 0);
```
0 is neither positive not negative. https://math.stackexchange.com/questions/26705/is-zero-positive-or-negative

vaibhavarora14 replied:
right, code is updated with 3dd02cd49ac897ea7001e3ebeb1253592843577a

ArfatSalman commented on js-exercises/sum-fibs/sumFibs.js:
```diff
@@ -0,0 +1,59 @@
+const isNumber = (arg) => typeof arg === 'number';
+
+const isOddNumber = (arg) => isNumber(arg) && (arg % 2 !== 0);
```
This would be true for `NaN`

vaibhavarora14 replied:
right, fixed issue with bcb3526112c874b7365fbbf8a957a7479e2f727c

ArfatSalman commented on js-exercises/sum-fibs/sumFibs.js:
```diff
@@ -0,0 +1,59 @@
+
+  const fibonacciIterator = fibonacciGenerator(startNumber);
+
+  for (const number of fibonacciIterator) {
```
You can also create a `filter` generator like and put the pass the isOddNumber there to only generate odd Fibonacci numbers.

vaibhavarora14 replied:
Maybe 8ab346c55474d57acbd63b0f97940c49cb53006b resolves the problem in context you pointed.
I am not sure, so please reply after viewing the changes 🙂

ArfatSalman commented on js-exercises/sum-fibs/sumFibs.js:
```diff
@@ -0,0 +1,59 @@
+
+  for (const number of fibonacciIterator) {
+    if (number > endNumber) {
+      break;
```
Can you do this without `break`?

vaibhavarora14 replied:
`break` is removed with 8ab346c55474d57acbd63b0f97940c49cb53006b

ArfatSalman reviewed (changes requested):
This is not a very good solution. You can do better by creating a Fibonacci number generator, a "filter" generator and pass the filtering condition to the `filter` function. Then you can only `take` n number of Fib numbers.

ArfatSalman commented:
Also, you can try using libraries like `lodash` to not write common utils again and again.

Commit:
Make code more loosely coupled
- Modify fibonacciGenerator to accept endNumber, as start number is always "1" in fibonacci series.
- Remove returnGenerator. It was added for some cool factor, but resulted in tight declaration of variables and assignment of values.
- Sum fibs function is cleaned with loose coupling of code, to handle any filter requirement easily if needed

Commit:
#5 Solve issue of isNumber getting true for isNaN

Commit:
#5 Fix mistake of isNumber done in prev commit

Commit:
#5 Zero is neither positive, nor negative. Fix issue in isPositiveNumber()

vaibhavarora14 commented:
@ArfatSalman, I am not going to use `loadash` very easily. 😀
I will write utilities, fix issue there, learn from them.
I am re-using those utilities by copy-paste.

Once code is in master, you will find utilities spread and re-used from code.
